iree/turbine/kernel/wave/codegen.py

```diff
@@ -1198,3 +1200,64 @@ def handle_get_result(emitter: WaveEmitter, node: fx.Node):
 @handle_op(operator.getitem)
 def handle_getitem(emitter: WaveEmitter, node: fx.Node):
     raise NotImplementedError("getitem: Currently only stub implementation")
+
+
+def get_float_type(bitwidth: int):
+    match bitwidth:
+        case 16:
+            return F16Type.get()
+        case 32:
+            return F32Type.get()
+        case _:
+            raise NotImplementedError(f"Unsupported float bitwidth: {bitwidth}")
+
+
+@handle_op(cast)
+def handle_cast(emitter: WaveEmitter, node: fx.Node):
+    try:
+        register, dtype = node.args
+    except ValueError as e:
+        raise ValidationError("Malformed arguments") from e
+    vector_src = cast_vector(emitter, register)
+    src_vector_type = vector_src.type
+    dst_elem_type = IrType.parse(dtype.ir_type_asm())
+    dst_vector_type = VectorType.get(src_vector_type.shape, dst_elem_type)
+
+    if src_vector_type == dst_vector_type:
+        emitter.bind_node_proxy(node, vector_src)
+        return
+
+    is_src_float = _is_float_type(src_vector_type.element_type)
+    is_dst_float = _is_float_type(dst_elem_type)
+
+    conversion_ops = {
+        (True, True): lambda _, x: x,
+        (False, False): lambda _, x: x,
+        (True, False): arith_d.fptosi,
+        (False, True): arith_d.sitofp,
+    }
+
+    cast_ops = {
+        (True, True): arith_d.extf,
+        (True, False): arith_d.extsi,
+        (False, True): arith_d.truncf,
+        (False, False): arith_d.trunci,
+    }
+
+    dtype = (
+        get_float_type(dst_elem_type.width)
+        if is_dst_float
+        else IntegerType.get_signless(dst_elem_type.width)
+    )
+    converted_vector = conversion_ops[(is_src_float, is_dst_float)](
+        VectorType.get(src_vector_type.shape, dtype), vector_src
```

Comment:
why can't we use dst_elem_type directly instead of `dtype`?

Reply:
nvm

Reply:
good point, I was under the impression that the bit width has to be the same for the conversion ops but that is not true. Fixed.